### PR TITLE
Address notification batch mode

### PR DESCRIPTION
**Pull request by beiwei30**
**What this PR does**:

Support address notification from registry center in batch mode, instead of omit from register center one by one. In this mode, we assume all events accumulated by registry center first, then registry center will notify the directory the complete list in batch mode.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: registry/directory/directory.go [hunk 1]
```diff
@@ -104,72 +104,129 @@ func (dir *RegistryDirectory) subscribe(url *common.URL) {
...
+// service event is passed in, then it is incremental mode, and if an array of service events are passed in, it is
+// batch mode, in this mode, we assume the registry center have the complete list of the service events, therefore
+// in this case, we can safely assume any cached invoker not in the incoming list can be removed. It is necessary
+// since in batch mode, the register center handles the different type of events by itself, then notify the directory
+// a batch of 'Update' events, instead of omit the different type of event one by one.
+func (dir *RegistryDirectory) refreshInvokers(events ...*registry.ServiceEvent) {
+	var oldInvokers []protocol.Invoker
+
+	// in batch mode, it is safe to remove since we have the complete list of events.
+	if events != nil && len(events) > 1 {
```
Comment (AlexStocks) [inline registry/directory/directory.go:131]: events != nil && len(events) > 1 ==> len(events) > 1

File: registry/protocol/protocol.go [hunk 2]
```diff
@@ -242,7 +242,12 @@ func newOverrideSubscribeListener(overriderUrl *common.URL, invoker protocol.Inv
 }
 
 // Notify will be triggered when a service change notification is received.
-func (nl *overrideSubscribeListener) Notify(event *registry.ServiceEvent) {
+func (nl *overrideSubscribeListener) Notify(events ...*registry.ServiceEvent) {
+	if events == nil || len(events) == 0 {
```
Comment (AlexStocks) [inline registry/protocol/protocol.go:246]: events == nil || len(events) == 0 ==> len(events) == 0

File: registry/directory/directory.go (hunk 1 above)
Comment (beiwei30) [reply registry/directory/directory.go]: done.

File: registry/protocol/protocol.go (hunk 2 above)
Comment (beiwei30) [reply registry/protocol/protocol.go]: done.

**Review by Patrick0308 — COMMENTED** [review]
Excellent, LGTM

File: registry/directory/directory.go [hunk 3]
```diff
@@ -104,72 +104,129 @@ func (dir *RegistryDirectory) subscribe(url *common.URL) {
...
 }
 
 // update the cacheServices and subscribe service from registry
-func (dir *RegistryDirectory) update(res *registry.ServiceEvent) {
-	if res == nil {
-		return
+func (dir *RegistryDirectory) update(events ...*registry.ServiceEvent) {
+	for _, event := range events {
+		logger.Debugf("registry update, result{%s}", event)
+		logger.Debugf("update service name: %s!", event.Service)
```
Comment (zouyx) [inline registry/directory/directory.go:115]: I think should move this block to line 142?

https://github.com/apache/dubbo-go/pull/741/files#diff-f1103e9118b5d7dcaaa8b1dac15314d7R142

File: registry/directory/directory.go [hunk 4]
```diff
@@ -104,72 +104,129 @@ func (dir *RegistryDirectory) subscribe(url *common.URL) {
...
+					oldInvokers = append(oldInvokers, invoker)
+				}
+			}
+			return true
+		})
+	}
 
+	for _, event := range events {
+		if oldInvoker, _ := dir.cacheInvokerByEvent(event); oldInvoker != nil {
+			oldInvokers = append(oldInvokers, oldInvoker)
```
Comment (zouyx) [inline registry/directory/directory.go:144]:
```suggestion
	for _, event := range events {
	        logger.Debugf("registry update, result{%s}", event)
		logger.Debugf("update service name: %s!", event.Service)
		if oldInvoker, _ := dir.cacheInvokerByEvent(event); oldInvoker != nil {
			oldInvokers = append(oldInvokers, oldInvoker)
```

**Review by Patrick0308 — APPROVED** [review]

File: registry/directory/directory.go (hunk 3 above)
Comment (beiwei30) [reply registry/directory/directory.go]: done.

File: registry/directory/directory.go (hunk 4 above)
Comment (beiwei30) [reply registry/directory/directory.go]: done.

**Review by AlexStocks — APPROVED** [review]

**Review by hxmhlt — APPROVED** [review]
LGTM

**Review by zouyx — APPROVED** [review]

**Pull request merged**